toolchain/check/facet_type.cpp

```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+    CARBON_CHECK(values_.size() < std::numeric_limits<int>::max());
+    auto new_index = static_cast<int>(values_.size());
+    values_.push_back(Value());
+    values_.back().index = new_index;
+    auto result = indexes_.Insert(ac, new_index);
+    CARBON_CHECK(result.is_inserted());
+    return values_.back();
+  }
+
+  // Maps to an index in both of vectors below. Constants with the same index
+  // value are considered equivalent.
+  Map<RewriteRule::AssociatedConstant, int> indexes_;
+  llvm::SmallVector<Value> values_;
+};
+
+// Returns the value of the `AssociatedConstantDecl` accessed by an
+// `ImplWitnessAccess`, or nullopt if the access is not to an associated
+// constant (such as for an access to an associated function, or an error).
+//
+// The AssociatedConstantDecl is mapped into the specific of the witness.
+static auto TryGetAssociatedConstantFromImplWitnessAccess(
+    Context& context, const SemIR::ImplWitnessAccess& witness_access)
+    -> std::optional<SemIR::ConstantId> {
+  auto witness = context.insts().GetAs<SemIR::LookupImplWitness>(
+      witness_access.witness_id);
+  auto specific_interface =
+      context.specific_interfaces().Get(witness.query_specific_interface_id);
+  const auto& interface =
+      context.interfaces().Get(specific_interface.interface_id);
+  auto assoc_entities =
+      context.inst_blocks().Get(interface.associated_entities_id);
+  auto decl_id = assoc_entities[witness_access.index.index];
+  if (decl_id == SemIR::ErrorInst::InstId) {
+    return std::nullopt;
+  }
+  if (!context.insts().Is<SemIR::AssociatedConstantDecl>(decl_id)) {
+    // Something in the interface other than an associated constant.
+    return std::nullopt;
+  }
+  auto decl_const_id = SemIR::GetConstantValueInSpecific(
+      context.sem_ir(), specific_interface.specific_id, decl_id);
+  return decl_const_id;
+}
+
+// Generates the rewrite rules retrieved from the witness tables found in a
+// FacetType.
+class RewriteRulesFromFacetType {
+ public:
+  explicit RewriteRulesFromFacetType(Context& context,
+                                     SemIR::FacetTypeId facet_type_id)
+      : context_(context) {
+    if (facet_type_id.has_value()) {
+      const auto& facet_type_info = context.facet_types().Get(facet_type_id);
+      it_ = facet_type_info.rewrite_constraints.begin();
+      end_ = facet_type_info.rewrite_constraints.end();
+    } else {
+      it_ = end_ = nullptr;
+    }
+  }
+
+  // Returns each RewriteRule found in the FacetType, then returns nullopt.
+  auto Next() -> std::optional<RewriteRule> {
+    while (it_ != end_) {
+      auto lhs_id = it_->lhs_id;
+      auto rhs_id = it_->rhs_id;
+      ++it_;
+
+      while (lhs_id == SemIR::ErrorInst::InstId ||
+             rhs_id == SemIR::ErrorInst::InstId) {
+        lhs_id = it_->lhs_id;
+        rhs_id = it_->rhs_id;
+        ++it_;
+        if (it_ == end_) {
+          return std::nullopt;
+        }
+      }
```

Comment:
I think the `it_ == end_` check should be before you read through `it_` rather than after, because in the outer loop `it_` is referring to the next constraint, not the current one.

With that changed, I think this can be simplified to:
```suggestion
      if (lhs_id == SemIR::ErrorInst::InstId ||
          rhs_id == SemIR::ErrorInst::InstId) {
        continue;
      }
```

Reply:
Oh, yes thanks good eye.